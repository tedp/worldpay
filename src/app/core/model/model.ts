export interface HypermediaLink {
  href: string;
}
export interface HypermediaCuries {
  href: string;
  name: string;
  templated: boolean;
}
export class HypermediaResult {
  outcome?: string;
  _links: {
    [s: string]: HypermediaLink | HypermediaCuries[]
  };
}

export interface Merchant {
  entity: string;
}

export interface Narrative {
  line1: string;
}

export interface Value {
  currency: string;
  amount: number;
}

export interface CardExpiryDate {
  month: number;
  year: number;
}

export interface PaymentInstrument {
  type: string;
  cardNumber: string;
  cardHolderName: string;
  cardExpiryDate: CardExpiryDate;
}

export interface Instruction {
  narrative: Narrative;
  value: Value;
  paymentInstrument: PaymentInstrument;
}

export interface PaymentRequest {
  transactionReference: string;
  merchant: Merchant;
  instruction: Instruction;
}
