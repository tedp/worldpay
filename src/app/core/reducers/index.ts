import { ActionReducerMap, MetaReducer } from '@ngrx/store';
import { environment } from '../../../environments/environment';
import { PaymentsState, paymentReducer } from './payments.reducer';

export interface AppState {
  readonly payments: PaymentsState;
}

export const reducers: ActionReducerMap<AppState> = {
  payments: paymentReducer
};

export const metaReducers: MetaReducer<AppState>[] = !environment.production ? [] : [];

export * from './payments.reducer';

