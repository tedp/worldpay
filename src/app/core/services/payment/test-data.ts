import { HypermediaResult, PaymentRequest, Instruction, PaymentInstrument, CardExpiryDate, Narrative, Value, Merchant } from '../../model/model';

export const queryResponse: HypermediaResult = {
    _links: {
      'payments:authorize': {
        href: 'https://try.access.worldpay.com/payments/authorizations'
      },
      'tokens:tokens': {
        href: 'https://try.access.worldpay.com/tokens'
      },
      curies: [
        {
          href: 'https://try.access.worldpay.com/rels/payments/{rel}',
          name: 'payments',
          templated: true
        },
        {
          href: 'https://try.access.worldpay.com/rels/tokens/{rel}.json',
          name: 'tokens',
          templated: true
        }
      ]
    }
  };

export const authorizationResponse: HypermediaResult = {
    outcome: 'authorized',
    _links: {
        'payments:cancel': {
            href: 'https://try.access.worldpay.com/payments/authorizations/cancellations/eyJrIjoiazNhYjYzMiJ9'
        },
        'payments:settle': {
            href: 'https://try.access.worldpay.com/payments/settlements/full/eyJrIjoiazNhYjYzMiJ9'
        },
        'payments:partialSettle': {
            href: 'https://try.access.worldpay.com/payments/settlements/partials/eyJrIjoiazNhYjYzMiJ9'
        },
        'payments:events': {
            href: 'https://try.access.worldpay.com/payments/events/eyJrIjoiazNhYjYzMiJ9'
        },
        curies: [
            {
                name: 'payments',
                href: 'https://try.access.worldpay.com/rels/payments/{rel}',
                templated: true
            }
        ]
    }
};

export const cancelResponse: HypermediaResult = {
  _links: {
      'payments:events': {
          href: 'https://try.access.worldpay.com/payments/events/eyJrIjoiazNhYjYzMiJ9'
      },
      curies: [
          {
              name: 'payments',
              href: 'https://try.access.worldpay.com/rels/payments/{rel}',
              templated: true
          }
      ]
  }
};

export const cardExpiryDate: CardExpiryDate = {
  month: 10,
  year: 2022
};

export const paymentInstrument: PaymentInstrument = {
  cardExpiryDate,
  cardHolderName: 'Bob Smith',
  cardNumber: '04584 34589 34534 34534',
  type: 'visa'
};

export const narrative: Narrative = {
  line1: 'Big Components LTD'
};

export const value: Value = {
  amount: 10,
  currency: 'GBP'
};

export const instruction: Instruction = {
  narrative,
  paymentInstrument,
  value
};

export const merchant: Merchant = {
  entity: 'FastPaymentService'
};

export const paymentRequest: PaymentRequest = {
  instruction,
  merchant,
  transactionReference: 'fjsnrj83md983'
};
