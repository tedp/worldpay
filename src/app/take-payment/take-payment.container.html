<form [formGroup]="instructionForm" (ngSubmit)="onSubmit()" novalidate>
  <mat-card class="shipping-card">
    <mat-card-header>
      <mat-card-title>Payment instruction</mat-card-title>
    </mat-card-header>
    
    <mat-card-content>
      <div formGroupName="instruction" novalidate>

        <div class="row">
          <div class="col">
            <mat-form-field class="full-width">
              <input matInput placeholder="Narrative" formControlName="narrative">
            </mat-form-field>
          </div>
        </div>
  
        <div formGroupName="value" novalidate>
          <div class="row">
            <div class="col">
              <mat-form-field class="full-width">
                <mat-select placeholder="Currency" formControlName="currency">
                  <mat-option *ngFor="let currency of currencies" [value]="currency">
                    {{ currency }}
                  </mat-option>
                </mat-select>
              </mat-form-field>
            </div>
            <div class="col">
              <mat-form-field class="full-width">
                <input matInput formControlName="amount">
              </mat-form-field>
            </div>
          </div>
        </div>

        <div formGroupName="paymentInstrument" novalidate>
          <div class="row">
            <div class="col">
              <mat-form-field class="full-width">
                <mat-select placeholder="Card type" formControlName="type">
                  <mat-option *ngFor="let cardType of cardTypes" [value]="cardType">
                    {{ cardType }}
                  </mat-option>
                </mat-select>
              </mat-form-field>
            </div>

            <div class="col">
              <mat-form-field class="full-width">
                <input matInput placeholder="Card number e.g. 67473 93828 28283 9832" formControlName="cardNumber">
              </mat-form-field>
            </div>
          </div>

          <div formGroupName="cardExpiryDate" novalidate>
            <div class="row">
              <div class="col">
                <mat-form-field class="full-width">
                  <input matInput placeholder="Expiry month e.g. 01" formControlName="month">
                </mat-form-field>
              </div>
              <div class="col">
                <mat-form-field class="full-width">
                  <input matInput placeholder="Expiry year e.g. 22" formControlName="year">
                </mat-form-field>
              </div>
            </div>
          </div>
        
          <div class="row">
            <div class="col">
              <mat-form-field class="full-width">
                <input matInput placeholder="Name as it appears on the card e.g. Mr D Bridges" formControlName="cardHolderName">
              </mat-form-field>
            </div>
          </div>
          
        </div>
      </div>
        
    </mat-card-content>
    <mat-card-actions>
      <button mat-raised-button type="submit" color="primary" [disabled]="instructionForm.pristine">Proceed</button>
      <button mat-raised-button type="reset" (click)="revert()" [disabled]="instructionForm.pristine">Revert</button>
    </mat-card-actions>
  </mat-card>
</form>
