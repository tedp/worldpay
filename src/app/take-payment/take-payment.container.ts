import { Component, OnInit } from '@angular/core';
import { FormControl, FormGroup, FormBuilder, Validators } from '@angular/forms';
import { PaymentInstrument, Instruction, PaymentRequest, Merchant } from '../core/model/model';
import { cloneDeep } from 'lodash';
import { Store } from '@ngrx/store';
import { AppState } from '../core/reducers';
import { Router } from '@angular/router';
import { AuthorizeAction } from '../core/actions/payments.actions';

@Component({
  selector: 'app-take-payment',
  templateUrl: './take-payment.container.html',
  styleUrls: ['./take-payment.container.scss']
})
export class TakePaymentContainer implements OnInit {

  instructionForm: FormGroup;
  name = new FormControl('');
  cardTypes = ['Visa', 'Mastercard', 'Amex'];
  currencies = ['GBP', 'USD', 'AUD'];
  private merchant: Merchant = {
    entity: 'ABC'
  };
  private transactionReference = 'XYZ';

  constructor(
    private formBuilder: FormBuilder,
    private store: Store<AppState>,
    private router: Router) {

    this.instructionForm = this.createFormGroupWithBuilder(formBuilder);
  }

  ngOnInit() {
  }

  onSubmit() {
    const instruction: Instruction = cloneDeep(this.instructionForm.value.instruction);

    const paymentRequest: PaymentRequest = {
      merchant: this.merchant,
      transactionReference: this.transactionReference,
      instruction
    };
    this.store.dispatch(new AuthorizeAction({ paymentRequest }));
    this.router.navigate(['/confirm-payment']);
  }

  revert() {
    this.instructionForm.reset();
  }

  private createFormGroupWithBuilder(formBuilder: FormBuilder) {
    return formBuilder.group({
      instruction: formBuilder.group({
        narrative: '',
        value: formBuilder.group({
          currency: [null, Validators.required],
          amount: [null, Validators.required]
        }),
        paymentInstrument: formBuilder.group({
          type: [null, Validators.required],
          cardNumber: [null, Validators.required],
          cardHolderName: [null, Validators.required],
          cardExpiryDate: formBuilder.group({
            month: '',
            year: ''
          })
        })
      })
    });
  }
}
