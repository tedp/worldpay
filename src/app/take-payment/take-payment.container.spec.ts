import { ComponentFixture, TestBed, fakeAsync } from '@angular/core/testing';

import { TakePaymentContainer } from './take-payment.container';
import { AppModule } from '../app.module';
import { APP_BASE_HREF } from '@angular/common';
import { Store } from '@ngrx/store';
import { of } from 'rxjs';
import { AppState } from '../core/reducers';
import { By } from '@angular/platform-browser';
import { AuthorizeAction } from '../core/actions/payments.actions';
import { authorizationResponse, paymentRequest, cancelResponse } from '../core/services/payment/test-data';
import { PaymentRequest } from '../core/model/model';
import { Router } from '@angular/router';


describe('TakePaymentComponent', () => {
  let component: TakePaymentContainer;
  let fixture: ComponentFixture<TakePaymentContainer>;
  let dispatchSpy;
  let navigateSpy;

  const store: Store<AppState> = null;


  function submitForm() {
    const submitButton = fixture.debugElement.query(By.css('[type="submit"]')).nativeElement;
    submitButton.click();
    fixture.detectChanges();
  }

  function populateForm() {
    enterInputElementvalue(paymentRequest.instruction.narrative.line1, 'narrative');
    enterInputElementvalue(paymentRequest.instruction.value.currency, 'currency');
    enterInputElementvalue(paymentRequest.instruction.value.amount, 'amount');
    enterInputElementvalue(paymentRequest.instruction.paymentInstrument.type, 'type');
    enterInputElementvalue(paymentRequest.instruction.paymentInstrument.cardNumber, 'cardNumber');
    enterInputElementvalue(paymentRequest.instruction.paymentInstrument.cardExpiryDate.month, 'month');
    enterInputElementvalue(paymentRequest.instruction.paymentInstrument.cardExpiryDate.year, 'year');
    enterInputElementvalue(paymentRequest.instruction.paymentInstrument.cardHolderName, 'cardHolderName');
    fixture.detectChanges();
  }

  function enterInputElementvalue(value, elementName: string) {
    const inputElement = fixture.debugElement.query(By.css(`[formControlName="${elementName}"]`)).nativeElement;
    inputElement.value = value;
    inputElement.dispatchEvent(new Event('input'));
  }

  beforeEach(fakeAsync(() => {
    dispatchSpy = jasmine.createSpy('dispatch');
    navigateSpy = jasmine.createSpy('router');

    TestBed.configureTestingModule({
      imports: [ AppModule ],
      providers:
      [
        { provide: Store, useValue: { dispatch: dispatchSpy } },
        { provide: APP_BASE_HREF, useValue : 'payment-cancelled' },
        { provide: Router, useValue: { navigate: navigateSpy } }
      ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TakePaymentContainer);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('#submit', () => {
    const expectedPaymentRequest = {
      merchant: {
        entity: 'ABC'
      },
      transactionReference: 'XYZ',
      instruction: {
        narrative: 'Big Components LTD',
        value: {
          currency: null,
          amount: '10'
        },
        paymentInstrument: {
          type: null,
          cardNumber: '04584 34589 34534 34534',
          cardHolderName: 'Bob Smith',
          cardExpiryDate: {
            month: '10',
            year: '2022'
          }
        }
      }
    } as PaymentRequest;

    const expectedAction = new AuthorizeAction({ paymentRequest: expectedPaymentRequest });

    populateForm();
    submitForm();
    expect(dispatchSpy.calls.mostRecent().args[0]).toEqual(expectedAction);
    expect(navigateSpy).toHaveBeenCalledWith(['/confirm-payment']);
  });
});
