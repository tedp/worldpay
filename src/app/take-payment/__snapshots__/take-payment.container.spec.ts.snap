// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TakePaymentComponent should match snapshot 1`] = `
<app-take-payment
  cardTypes={[Function Array]}
  currencies={[Function Array]}
  formBuilder={[Function FormBuilder]}
  instructionForm={[Function FormGroup]}
  merchant={[Function Object]}
  name={[Function FormControl]}
  router={[Function Object]}
  store={[Function Object]}
  transactionReference={[Function String]}
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <mat-card
      class="shipping-card mat-card"
    >
      <mat-card-header
        class="mat-card-header"
      >
        <div
          class="mat-card-header-text"
        >
          <mat-card-title
            class="mat-card-title"
          >
            Payment instruction
          </mat-card-title>
        </div>
      </mat-card-header>
      <mat-card-content
        class="mat-card-content"
      >
        <div
          class="ng-untouched ng-pristine ng-invalid"
          formgroupname="instruction"
          ng-reflect-name="instruction"
          novalidate=""
        >
          <div
            class="row"
          >
            <div
              class="col"
            >
              <mat-form-field
                class="full-width mat-form-field ng-tns-c1-0 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-valid _mat-animation-noopable"
              >
                <div
                  class="mat-form-field-wrapper"
                >
                  <div
                    class="mat-form-field-flex"
                  >
                    
                    
                    <div
                      class="mat-form-field-infix"
                    >
                      <input
                        aria-invalid="false"
                        aria-required="false"
                        class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-valid"
                        formcontrolname="narrative"
                        id="mat-input-0"
                        matinput=""
                        ng-reflect-name="narrative"
                        ng-reflect-placeholder="Narrative"
                        placeholder="Narrative"
                      />
                      <span
                        class="mat-form-field-label-wrapper"
                      >
                        
                        <label
                          aria-owns="mat-input-0"
                          class="mat-form-field-label ng-tns-c1-0 mat-empty mat-form-field-empty ng-star-inserted"
                          for="mat-input-0"
                          id="mat-form-field-label-1"
                          ng-reflect-disabled="true"
                          ng-reflect-ng-switch="false"
                        >
                          
                          
                          Narrative
                          
                          
                        </label>
                      </span>
                    </div>
                    
                  </div>
                  
                  <div
                    class="mat-form-field-underline ng-tns-c1-0 ng-star-inserted"
                  >
                    <span
                      class="mat-form-field-ripple"
                    />
                  </div>
                  <div
                    class="mat-form-field-subscript-wrapper"
                    ng-reflect-ng-switch="hint"
                  >
                    
                    
                    <div
                      class="mat-form-field-hint-wrapper ng-tns-c1-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                      style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                    >
                      
                      <div
                        class="mat-form-field-hint-spacer"
                      />
                    </div>
                  </div>
                </div>
              </mat-form-field>
            </div>
          </div>
          <div
            class="ng-untouched ng-pristine ng-invalid"
            formgroupname="value"
            ng-reflect-name="value"
            novalidate=""
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <mat-form-field
                  class="full-width mat-form-field ng-tns-c1-1 mat-primary mat-form-field-type-mat-select mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid _mat-animation-noopable"
                >
                  <div
                    class="mat-form-field-wrapper"
                  >
                    <div
                      class="mat-form-field-flex"
                    >
                      
                      
                      <div
                        class="mat-form-field-infix"
                      >
                        <mat-select
                          aria-disabled="false"
                          aria-invalid="false"
                          aria-label="Currency"
                          aria-multiselectable="false"
                          aria-required="false"
                          class="mat-select ng-tns-c2-2 ng-untouched ng-pristine ng-invalid mat-select-empty ng-star-inserted"
                          formcontrolname="currency"
                          id="mat-select-0"
                          ng-reflect-name="currency"
                          ng-reflect-placeholder="Currency"
                          placeholder="Currency"
                          role="listbox"
                          tabindex="0"
                        >
                          <div
                            aria-hidden="true"
                            cdk-overlay-origin=""
                            class="mat-select-trigger"
                          >
                            <div
                              class="mat-select-value"
                              ng-reflect-ng-switch="true"
                            >
                              
                              <span
                                class="mat-select-placeholder ng-tns-c2-2 ng-star-inserted"
                              >
                                Currency
                              </span>
                              
                            </div>
                            <div
                              class="mat-select-arrow-wrapper"
                            >
                              <div
                                class="mat-select-arrow"
                              />
                            </div>
                          </div>
                          
                        </mat-select>
                        <span
                          class="mat-form-field-label-wrapper"
                        >
                          
                          <label
                            aria-owns="mat-select-0"
                            class="mat-form-field-label ng-tns-c1-1 mat-empty mat-form-field-empty ng-star-inserted"
                            for="mat-select-0"
                            id="mat-form-field-label-3"
                            ng-reflect-disabled="true"
                            ng-reflect-ng-switch="false"
                          >
                            
                            
                            Currency
                            
                            
                          </label>
                        </span>
                      </div>
                      
                    </div>
                    
                    <div
                      class="mat-form-field-underline ng-tns-c1-1 ng-star-inserted"
                    >
                      <span
                        class="mat-form-field-ripple"
                      />
                    </div>
                    <div
                      class="mat-form-field-subscript-wrapper"
                      ng-reflect-ng-switch="hint"
                    >
                      
                      
                      <div
                        class="mat-form-field-hint-wrapper ng-tns-c1-1 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                        style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                      >
                        
                        <div
                          class="mat-form-field-hint-spacer"
                        />
                      </div>
                    </div>
                  </div>
                </mat-form-field>
              </div>
              <div
                class="col"
              >
                <mat-form-field
                  class="full-width mat-form-field ng-tns-c1-3 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid _mat-animation-noopable"
                >
                  <div
                    class="mat-form-field-wrapper"
                  >
                    <div
                      class="mat-form-field-flex"
                    >
                      
                      
                      <div
                        class="mat-form-field-infix"
                      >
                        <input
                          aria-invalid="false"
                          aria-required="false"
                          class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-invalid"
                          formcontrolname="amount"
                          id="mat-input-1"
                          matinput=""
                          ng-reflect-name="amount"
                        />
                        <span
                          class="mat-form-field-label-wrapper"
                        >
                          
                        </span>
                      </div>
                      
                    </div>
                    
                    <div
                      class="mat-form-field-underline ng-tns-c1-3 ng-star-inserted"
                    >
                      <span
                        class="mat-form-field-ripple"
                      />
                    </div>
                    <div
                      class="mat-form-field-subscript-wrapper"
                      ng-reflect-ng-switch="hint"
                    >
                      
                      
                      <div
                        class="mat-form-field-hint-wrapper ng-tns-c1-3 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                        style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                      >
                        
                        <div
                          class="mat-form-field-hint-spacer"
                        />
                      </div>
                    </div>
                  </div>
                </mat-form-field>
              </div>
            </div>
          </div>
          <div
            class="ng-untouched ng-pristine ng-invalid"
            formgroupname="paymentInstrument"
            ng-reflect-name="paymentInstrument"
            novalidate=""
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <mat-form-field
                  class="full-width mat-form-field ng-tns-c1-4 mat-primary mat-form-field-type-mat-select mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid _mat-animation-noopable"
                >
                  <div
                    class="mat-form-field-wrapper"
                  >
                    <div
                      class="mat-form-field-flex"
                    >
                      
                      
                      <div
                        class="mat-form-field-infix"
                      >
                        <mat-select
                          aria-disabled="false"
                          aria-invalid="false"
                          aria-label="Card type"
                          aria-multiselectable="false"
                          aria-required="false"
                          class="mat-select ng-tns-c2-5 ng-untouched ng-pristine ng-invalid mat-select-empty ng-star-inserted"
                          formcontrolname="type"
                          id="mat-select-1"
                          ng-reflect-name="type"
                          ng-reflect-placeholder="Card type"
                          placeholder="Card type"
                          role="listbox"
                          tabindex="0"
                        >
                          <div
                            aria-hidden="true"
                            cdk-overlay-origin=""
                            class="mat-select-trigger"
                          >
                            <div
                              class="mat-select-value"
                              ng-reflect-ng-switch="true"
                            >
                              
                              <span
                                class="mat-select-placeholder ng-tns-c2-5 ng-star-inserted"
                              >
                                Card type
                              </span>
                              
                            </div>
                            <div
                              class="mat-select-arrow-wrapper"
                            >
                              <div
                                class="mat-select-arrow"
                              />
                            </div>
                          </div>
                          
                        </mat-select>
                        <span
                          class="mat-form-field-label-wrapper"
                        >
                          
                          <label
                            aria-owns="mat-select-1"
                            class="mat-form-field-label ng-tns-c1-4 mat-empty mat-form-field-empty ng-star-inserted"
                            for="mat-select-1"
                            id="mat-form-field-label-7"
                            ng-reflect-disabled="true"
                            ng-reflect-ng-switch="false"
                          >
                            
                            
                            Card type
                            
                            
                          </label>
                        </span>
                      </div>
                      
                    </div>
                    
                    <div
                      class="mat-form-field-underline ng-tns-c1-4 ng-star-inserted"
                    >
                      <span
                        class="mat-form-field-ripple"
                      />
                    </div>
                    <div
                      class="mat-form-field-subscript-wrapper"
                      ng-reflect-ng-switch="hint"
                    >
                      
                      
                      <div
                        class="mat-form-field-hint-wrapper ng-tns-c1-4 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                        style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                      >
                        
                        <div
                          class="mat-form-field-hint-spacer"
                        />
                      </div>
                    </div>
                  </div>
                </mat-form-field>
              </div>
              <div
                class="col"
              >
                <mat-form-field
                  class="full-width mat-form-field ng-tns-c1-6 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid _mat-animation-noopable"
                >
                  <div
                    class="mat-form-field-wrapper"
                  >
                    <div
                      class="mat-form-field-flex"
                    >
                      
                      
                      <div
                        class="mat-form-field-infix"
                      >
                        <input
                          aria-invalid="false"
                          aria-required="false"
                          class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-invalid"
                          formcontrolname="cardNumber"
                          id="mat-input-2"
                          matinput=""
                          ng-reflect-name="cardNumber"
                          ng-reflect-placeholder="Card number e.g. 67473 93828 2"
                          placeholder="Card number e.g. 67473 93828 28283 9832"
                        />
                        <span
                          class="mat-form-field-label-wrapper"
                        >
                          
                          <label
                            aria-owns="mat-input-2"
                            class="mat-form-field-label ng-tns-c1-6 mat-empty mat-form-field-empty ng-star-inserted"
                            for="mat-input-2"
                            id="mat-form-field-label-9"
                            ng-reflect-disabled="true"
                            ng-reflect-ng-switch="false"
                          >
                            
                            
                            Card number e.g. 67473 93828 28283 9832
                            
                            
                          </label>
                        </span>
                      </div>
                      
                    </div>
                    
                    <div
                      class="mat-form-field-underline ng-tns-c1-6 ng-star-inserted"
                    >
                      <span
                        class="mat-form-field-ripple"
                      />
                    </div>
                    <div
                      class="mat-form-field-subscript-wrapper"
                      ng-reflect-ng-switch="hint"
                    >
                      
                      
                      <div
                        class="mat-form-field-hint-wrapper ng-tns-c1-6 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                        style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                      >
                        
                        <div
                          class="mat-form-field-hint-spacer"
                        />
                      </div>
                    </div>
                  </div>
                </mat-form-field>
              </div>
            </div>
            <div
              class="ng-untouched ng-pristine ng-valid"
              formgroupname="cardExpiryDate"
              ng-reflect-name="cardExpiryDate"
              novalidate=""
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  <mat-form-field
                    class="full-width mat-form-field ng-tns-c1-7 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-valid _mat-animation-noopable"
                  >
                    <div
                      class="mat-form-field-wrapper"
                    >
                      <div
                        class="mat-form-field-flex"
                      >
                        
                        
                        <div
                          class="mat-form-field-infix"
                        >
                          <input
                            aria-invalid="false"
                            aria-required="false"
                            class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-valid"
                            formcontrolname="month"
                            id="mat-input-3"
                            matinput=""
                            ng-reflect-name="month"
                            ng-reflect-placeholder="Expiry month e.g. 01"
                            placeholder="Expiry month e.g. 01"
                          />
                          <span
                            class="mat-form-field-label-wrapper"
                          >
                            
                            <label
                              aria-owns="mat-input-3"
                              class="mat-form-field-label ng-tns-c1-7 mat-empty mat-form-field-empty ng-star-inserted"
                              for="mat-input-3"
                              id="mat-form-field-label-11"
                              ng-reflect-disabled="true"
                              ng-reflect-ng-switch="false"
                            >
                              
                              
                              Expiry month e.g. 01
                              
                              
                            </label>
                          </span>
                        </div>
                        
                      </div>
                      
                      <div
                        class="mat-form-field-underline ng-tns-c1-7 ng-star-inserted"
                      >
                        <span
                          class="mat-form-field-ripple"
                        />
                      </div>
                      <div
                        class="mat-form-field-subscript-wrapper"
                        ng-reflect-ng-switch="hint"
                      >
                        
                        
                        <div
                          class="mat-form-field-hint-wrapper ng-tns-c1-7 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                          style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                        >
                          
                          <div
                            class="mat-form-field-hint-spacer"
                          />
                        </div>
                      </div>
                    </div>
                  </mat-form-field>
                </div>
                <div
                  class="col"
                >
                  <mat-form-field
                    class="full-width mat-form-field ng-tns-c1-8 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-valid _mat-animation-noopable"
                  >
                    <div
                      class="mat-form-field-wrapper"
                    >
                      <div
                        class="mat-form-field-flex"
                      >
                        
                        
                        <div
                          class="mat-form-field-infix"
                        >
                          <input
                            aria-invalid="false"
                            aria-required="false"
                            class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-valid"
                            formcontrolname="year"
                            id="mat-input-4"
                            matinput=""
                            ng-reflect-name="year"
                            ng-reflect-placeholder="Expiry year e.g. 22"
                            placeholder="Expiry year e.g. 22"
                          />
                          <span
                            class="mat-form-field-label-wrapper"
                          >
                            
                            <label
                              aria-owns="mat-input-4"
                              class="mat-form-field-label ng-tns-c1-8 mat-empty mat-form-field-empty ng-star-inserted"
                              for="mat-input-4"
                              id="mat-form-field-label-13"
                              ng-reflect-disabled="true"
                              ng-reflect-ng-switch="false"
                            >
                              
                              
                              Expiry year e.g. 22
                              
                              
                            </label>
                          </span>
                        </div>
                        
                      </div>
                      
                      <div
                        class="mat-form-field-underline ng-tns-c1-8 ng-star-inserted"
                      >
                        <span
                          class="mat-form-field-ripple"
                        />
                      </div>
                      <div
                        class="mat-form-field-subscript-wrapper"
                        ng-reflect-ng-switch="hint"
                      >
                        
                        
                        <div
                          class="mat-form-field-hint-wrapper ng-tns-c1-8 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                          style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                        >
                          
                          <div
                            class="mat-form-field-hint-spacer"
                          />
                        </div>
                      </div>
                    </div>
                  </mat-form-field>
                </div>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <mat-form-field
                  class="full-width mat-form-field ng-tns-c1-9 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid _mat-animation-noopable"
                >
                  <div
                    class="mat-form-field-wrapper"
                  >
                    <div
                      class="mat-form-field-flex"
                    >
                      
                      
                      <div
                        class="mat-form-field-infix"
                      >
                        <input
                          aria-invalid="false"
                          aria-required="false"
                          class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-invalid"
                          formcontrolname="cardHolderName"
                          id="mat-input-5"
                          matinput=""
                          ng-reflect-name="cardHolderName"
                          ng-reflect-placeholder="Name as it appears on the card"
                          placeholder="Name as it appears on the card e.g. Mr D Bridges"
                        />
                        <span
                          class="mat-form-field-label-wrapper"
                        >
                          
                          <label
                            aria-owns="mat-input-5"
                            class="mat-form-field-label ng-tns-c1-9 mat-empty mat-form-field-empty ng-star-inserted"
                            for="mat-input-5"
                            id="mat-form-field-label-15"
                            ng-reflect-disabled="true"
                            ng-reflect-ng-switch="false"
                          >
                            
                            
                            Name as it appears on the card e.g. Mr D Bridges
                            
                            
                          </label>
                        </span>
                      </div>
                      
                    </div>
                    
                    <div
                      class="mat-form-field-underline ng-tns-c1-9 ng-star-inserted"
                    >
                      <span
                        class="mat-form-field-ripple"
                      />
                    </div>
                    <div
                      class="mat-form-field-subscript-wrapper"
                      ng-reflect-ng-switch="hint"
                    >
                      
                      
                      <div
                        class="mat-form-field-hint-wrapper ng-tns-c1-9 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                        style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
                      >
                        
                        <div
                          class="mat-form-field-hint-spacer"
                        />
                      </div>
                    </div>
                  </div>
                </mat-form-field>
              </div>
            </div>
          </div>
        </div>
      </mat-card-content>
      <mat-card-actions
        class="mat-card-actions"
      >
        <button
          class="mat-raised-button mat-primary _mat-animation-noopable"
          color="primary"
          disabled=""
          mat-raised-button=""
          ng-reflect-color="primary"
          ng-reflect-disabled="true"
          type="submit"
        >
          <span
            class="mat-button-wrapper"
          >
            Proceed
          </span>
          <div
            class="mat-button-ripple mat-ripple"
            matripple=""
            ng-reflect-centered="false"
            ng-reflect-disabled="true"
            ng-reflect-trigger="[object HTMLButtonElement]"
          />
          <div
            class="mat-button-focus-overlay"
          />
        </button>
        <button
          class="mat-raised-button _mat-animation-noopable"
          disabled=""
          mat-raised-button=""
          ng-reflect-disabled="true"
          type="reset"
        >
          <span
            class="mat-button-wrapper"
          >
            Revert
          </span>
          <div
            class="mat-button-ripple mat-ripple"
            matripple=""
            ng-reflect-centered="false"
            ng-reflect-disabled="true"
            ng-reflect-trigger="[object HTMLButtonElement]"
          />
          <div
            class="mat-button-focus-overlay"
          />
        </button>
      </mat-card-actions>
    </mat-card>
  </form>
</app-take-payment>
`;
